chcemvediet/apps/anonymization/cron.py

```diff
@@ -14,23 +21,65 @@ def attachment_normalization():
             .filter(attachmentnormalization__isnull=True)
             .first()
             )
-
     if attachment is None:
         return
-    if attachment.content_type == content_types.PDF_CONTENT_TYPE:
-        AttachmentNormalization.objects.create(
-            attachment=attachment,
-            successful=True,
-            file=ContentFile(attachment.content),
-            content_type=content_types.PDF_CONTENT_TYPE
-        )
-        cron_logger.info(u'Normalized attachment: {}'.format(attachment))
+    elif attachment.content_type == content_types.PDF_CONTENT_TYPE:
+        normalize_pdf(attachment)
+    elif attachment.content_type in content_types.LIBREOFFICE_CONTENT_TYPES:
+        normalize_using_libreoffice(attachment)
     else:
+        skip_normalization(attachment)
+
+def normalize_pdf(attachment):
+    AttachmentNormalization.objects.create(
+        attachment=attachment,
+        successful=True,
+        file=ContentFile(attachment.content),
+        content_type=content_types.PDF_CONTENT_TYPE
+    )
+    cron_logger.info(u'Normalized attachment: {}'.format(attachment))
+
+def normalize_using_libreoffice(attachment):
+    directory_name = tempfile.mkdtemp()
+    try:
+        filename = os.path.join(directory_name, u'file' + guess_extension(attachment.content_type))
+        shutil.copy2(attachment.file.path, filename)
+        p = subprocess32.run(
+            [u'libreoffice', u'--headless', u'--convert-to', u'pdf', u'--outdir', directory_name,
+             filename],
+            stdout=subprocess32.PIPE,
+            stderr=subprocess32.PIPE,
+            timeout=300
+        )
+        with open(os.path.join(directory_name, u'file.pdf'), u'rb') as file_pdf:
+            AttachmentNormalization.objects.create(
+                attachment=attachment,
+                successful=True,
+                file=ContentFile(file_pdf.read()),
+                content_type=content_types.PDF_CONTENT_TYPE,
+                debug=u'STDOUT:\n{}\nSTDERR:\n{}'.format(p.stdout, p.stderr)
+            )
+        cron_logger.info(u'Normalized attachment using libreoffice: {}'.format(attachment))
+    except Exception as result:
+        trace = unicode(traceback.format_exc(), u'utf-8')
         AttachmentNormalization.objects.create(
             attachment=attachment,
             successful=False,
-            content_type=None,
-            debug=u'Not supported content type'
+            content_type=content_types.PDF_CONTENT_TYPE,
+            debug=u'{}'.format(trace)
```

Comment:
Pls aj pri neuspechu uloz stdout a stderr do `debug`. V pythone nema `try` samostatny scope, takze premenna `p` prezila a mozes ju tu pouzit.

Akurat pozor na to, ze exception mohla nastat pred, pocas alebo po vykonani `subprocess32.run` a treba osetrit vsetky pripady. Ak nastala **pred**, tak `p` je `None` a zalogovat netreba nic. Ak nastala **po** tak `p` je definovane a da sa z neho precitat `p.stdout` a `p.stderr`. Ak nastala **pocas**, tak `p` je stale `None`, ale v tom pripade `subprocess32.run` raisol exception [`TimeoutExpired`] alebo [`CalledProcessError`], ktore obe maju `e.stdout` a `e.stderr` atributy.

Cize nieco ako:
```
stdout = p.stdout if p else getattr(e, 'stdout', None)
```

[`TimeoutExpired`]: https://docs.python.org/3.6/library/subprocess.html#subprocess.TimeoutExpired
[`CalledProcessError`]: https://docs.python.org/3.6/library/subprocess.html#subprocess.CalledProcessError

Reply:
:ok:

Reply:
Ano, mas pravdu.